File: backend/app/schemas/memory.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel

# ── MemoryDump ─────────────────────────────────────────────────────────────────

class MemoryDumpOut(BaseModel):
    id:           str
    case_id:      str
    filename:     str
    os_type:      str                  # "windows" | "linux"
    file_size:    int | None
    status:       str                  # uploaded | analyzing | done | error
    error_msg:    str | None
    uploaded_at:  datetime

    model_config = {"from_attributes": True}


# ── MemoryPluginResult ─────────────────────────────────────────────────────────

class MemoryPluginResultOut(BaseModel):
    id:           int
    dump_id:      str
    plugin_name:  str
    plugin_args:  dict[str, Any] | None
    status:       str                  # pending | running | done | error
    output:       str | None
    error:        str | None
    started_at:   datetime | None
    completed_at: datetime | None
    is_custom:    bool

    model_config = {"from_attributes": True}


# ── Run custom plugin payload ──────────────────────────────────────────────────

class RunPluginPayload(BaseModel):
    plugin_name: str                   # e.g. "windows.pslist"
    plugin_args: dict[str, Any] | None = None   # {pid: 1234, ...}

File: frontend/src/api/memory.ts
import api from './client'

export interface MemoryDump {
  id:          string
  case_id:     string
  filename:    string
  os_type:     string    // "windows" | "linux"
  file_size:   number | null
  status:      string    // uploaded | analyzing | done | error
  error_msg:   string | null
  uploaded_at: string
}

export interface MemoryPluginResult {
  id:           number
  dump_id:      string
  plugin_name:  string
  plugin_args:  Record<string, unknown> | null
  status:       string   // pending | running | done | error
  output:       string | null
  error:        string | null
  started_at:   string | null
  completed_at: string | null
  is_custom:    boolean
}

export const memoryApi = {
  upload: (caseId: string, formData: FormData): Promise<MemoryDump> =>
    api.post<MemoryDump>(`/memory/${caseId}/upload`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    }).then(r => r.data),

  listDumps: (caseId: string): Promise<MemoryDump[]> =>
    api.get<MemoryDump[]>(`/memory/${caseId}/dumps`).then(r => r.data),

  getDump: (caseId: string, dumpId: string): Promise<MemoryDump> =>
    api.get<MemoryDump>(`/memory/${caseId}/dumps/${dumpId}`).then(r => r.data),

  deleteDump: (caseId: string, dumpId: string): Promise<void> =>
    api.delete(`/memory/${caseId}/dumps/${dumpId}`).then(() => undefined),

  listPlugins: (caseId: string, dumpId: string): Promise<MemoryPluginResult[]> =>
    api.get<MemoryPluginResult[]>(`/memory/${caseId}/dumps/${dumpId}/plugins`).then(r => r.data),

  runPlugin: (
    caseId: string,
    dumpId: string,
    plugin_name: string,
    plugin_args?: Record<string, unknown>,
  ): Promise<MemoryPluginResult> =>
    api.post<MemoryPluginResult>(`/memory/${caseId}/dumps/${dumpId}/run`, {
      plugin_name,
      plugin_args: plugin_args ?? null,
    }).then(r => r.data),

  rerunPlugin: (caseId: string, dumpId: string, pluginId: number): Promise<MemoryPluginResult> =>
    api.post<MemoryPluginResult>(
      `/memory/${caseId}/dumps/${dumpId}/plugins/${pluginId}/rerun`, {}
    ).then(r => r.data),
}
